Tool for sharding fastqs

tfenne opened the pull request.

tfenne committed:
First working version of a tool to shard fastqs.

tfenne committed:
First working version of a tool to shard fastqs.

tfenne committed:
Added read-ahead on the input FASTQs.

tfenne committed:
Added read-ahead on the input FASTQs.

tfenne committed:
Cleanup and add docs.

tfenne committed:
Cleanup and add docs.

tfenne committed:
Added some basic tests.

tfenne committed:
Added some basic tests.

clintval commented on src/bin/commands/shard.rs:
```diff
@@ -0,0 +1,368 @@
+
+/// Struct to hold all the output writers for a single shard
+struct ShardWriters<W: Write> {
+    /// Name of the sample this set of writers is for
```
Stale doc? It is more like an ID than it is a name.


```suggestion
    /// ID of the shard this set of writers is for
```

clintval reviewed (commented):
Looks like Rust I would have written, which may be a terrible thing, or a great thing! 

Would you humor me and allow me to install CodeRabbit into this repo so I can see what that AI would say?

tfenne commented:
> Looks like Rust I would have written, which may be a terrible thing, or a great thing!
> 
> Would you humor me and allow me to install CodeRabbit into this repo so I can see what that AI would say?

Go for it!

tfenne committed:
Address review feedback from Clint and CodeRabbit.

- Fix error message in ShardWriters::write to report writers.len() rather
  than shard_number.
- Replace the FASTQ-source-out-of-sync assert_eq! with a propagated
  anyhow error so length mismatches in user-supplied inputs are reported
  cleanly instead of panicking.
- Lift the per-iteration record Vec allocation out of the main loop and
  reuse it via clear().
- Clarify that --threads controls output-compression threads.
- Correct the shard_number doc comment to refer to the shard ID.
- Fix the multi-file sharding test, which previously joined r1.fq twice
  and so wasn't actually exercising multiple inputs.
- Misc clippy/style cleanups in the new test code.

tfenne committed:
Address review feedback from Clint and CodeRabbit.

- Fix error message in ShardWriters::write to report writers.len() rather
  than shard_number.
- Replace the FASTQ-source-out-of-sync assert_eq! with a propagated
  anyhow error so length mismatches in user-supplied inputs are reported
  cleanly instead of panicking.
- Lift the per-iteration record Vec allocation out of the main loop and
  reuse it via clear().
- Clarify that --threads controls output-compression threads.
- Correct the shard_number doc comment to refer to the shard ID.
- Fix the multi-file sharding test, which previously joined r1.fq twice
  and so wasn't actually exercising multiple inputs.
- Misc clippy/style cleanups in the new test code.

tfenne committed:
Expand shard test coverage and address review feedback.

Tests:
- Adds 11 new tests in src/bin/commands/shard.rs covering round-robin
  assignment correctness, paired-record alignment, no-loss/no-duplication,
  single-shard / more-shards-than-reads / empty-input edge cases,
  mismatched-input error path, gzipped inputs, BGZF inputs spanning
  multiple BGZF blocks (verifies we don't stop at the first block), and
  that outputs round-trip cleanly through the bgzf crate's Reader.
- Refactors test helpers: shared write_fastq_records core, parameterised
  build_sharder / collect_shard_outputs, and a new build_fastq_bgzf
  helper that uses bgzf::Writer.  Adds bgzf as a dev-dependency.
- Fixes a latent bug in build_fastq's name format: doc said
  "{prefix}{idx}{suffix}" but code emitted "{prefix}{suffix}{idx}",
  which made /1 and /2 read-end markers run into the index.

Review feedback (CodeRabbit, second pass):
- Adds clap range validators to --shards (>=1) and --threads (>=2) so
  zero values are rejected at parse time rather than panicking via
  shard_writers[0] / `% 0` / queue_size(0).
- Reserves one thread for the main loop in build_writer_pool by passing
  self.threads - 1 to the pool, matching the pattern in subsample.rs.
- Doc nit: "uncompressed, gzipped, or block-gzipped" (missing comma).

tfenne committed:
Expand shard test coverage and address review feedback.

Tests:
- Adds 11 new tests in src/bin/commands/shard.rs covering round-robin
  assignment correctness, paired-record alignment, no-loss/no-duplication,
  single-shard / more-shards-than-reads / empty-input edge cases,
  mismatched-input error path, gzipped inputs, BGZF inputs spanning
  multiple BGZF blocks (verifies we don't stop at the first block), and
  that outputs round-trip cleanly through the bgzf crate's Reader.
- Refactors test helpers: shared write_fastq_records core, parameterised
  build_sharder / collect_shard_outputs, and a new build_fastq_bgzf
  helper that uses bgzf::Writer.  Adds bgzf as a dev-dependency.
- Fixes a latent bug in build_fastq's name format: doc said
  "{prefix}{idx}{suffix}" but code emitted "{prefix}{suffix}{idx}",
  which made /1 and /2 read-end markers run into the index.

Review feedback (CodeRabbit, second pass):
- Adds clap range validators to --shards (>=1) and --threads (>=2) so
  zero values are rejected at parse time rather than panicking via
  shard_writers[0] / `% 0` / queue_size(0).
- Reserves one thread for the main loop in build_writer_pool by passing
  self.threads - 1 to the pool, matching the pattern in subsample.rs.
- Doc nit: "uncompressed, gzipped, or block-gzipped" (missing comma).

nh13 commented:
@coderabbitai review

tfenne committed:
Address review feedback and bump input buffer to 1 MiB.

- Use try_for_each instead of collecting into a throwaway Vec when closing
  pooled writers, matching the pattern in demux.rs.
- Reword the into_parts doc comment for clarity.
- Bump the input read buffer from 64 KiB to 1 MiB. Benchmarked on a 5 GB
  BGZF input sharded 4 ways with 8 threads: 1 MiB is a consistent ~3.4%
  faster on wall time (57.4s vs 59.5s) with identical CPU, and it matches
  the buffer size already used by demux and subsample.

tfenne committed:
Fix tiny FastqReader buffer in shard input readers.

The input FastqReaders were constructed with a 10-byte capacity, forcing
seq_io to refill its buffer almost per-record. Use BUFFER_SIZE (1 MiB) to
match demux. Benchmarked on a 5 GB single-input shard (4 shards, 8 threads):
~2% faster wall and notably tighter run-to-run variance (sigma 1.14s -> 0.27s).

tfenne committed:
Switch flate2 input decompression to the zlib-rs backend.

Replaces the zlib-ng (C) backend with the pure-Rust zlib-rs for gzip input
decoding across the toolkit (demux, shard, subsample). Profiling showed the
single-input shard path is bottlenecked on the single decompress+parse thread,
with gzip inflate the largest slice. Benchmarked on a 5 GB single-input shard
(4 shards, 8 threads) on Apple Silicon: ~6% faster wall (54.7s vs 58.4s) with
slightly lower CPU. Also drops flate2's C-compiler build requirement.

Note: benchmarked only on aarch64 so far; relative performance vs zlib-ng on
x86_64 still to be confirmed.

tfenne committed:
Parse FASTQ input on the main thread; default shard compression to 1.

Splits gzip decompression and FASTQ parsing across threads.  Each input's
background read-ahead thread now decompresses into fixed-size byte chunks
(ByteChunker), which the main thread parses into borrowed RefRecords (via a
ChunkReader Read adapter) and emits verbatim with write_unchanged.  This keeps
inflation off the main thread and removes the per-record OwnedRecord allocation.

On a 5 GB single-input shard (4 shards, 8 threads, Apple Silicon) this is 1.30x
faster (39.3s vs 51.1s).  Profiling confirms the bottleneck moved from the
single decompress+parse thread to the compression pool: compressors ~94% busy,
main thread now idling on compression backpressure rather than input.

Adds hidden --chunk-size/--chunk-count knobs to tune the decompressed chunk
buffers without rebuilding.

Also lowers the default output compression level from 5 to 1.  Sharded FASTQs
are typically short-lived intermediates where write throughput matters more than
file size; benchmarking showed level 1 trades ~7% larger output for roughly half
the compression CPU at similar wall time (paired-end gains more, since two
decompression threads raise the input ceiling).

tfenne committed:
Address review feedback on shard.

- Restore validate-before-write in the sharding loop: pull the full record set
  from every input and check for an out-of-sync (short) set before writing any
  of it, so a mismatched-length input no longer leaves a misaligned orphan
  record in the output files before the run errors.
- Validate --compression-level as 1..=12 at the CLI so bad values are rejected
  up front instead of failing late after the read-ahead/compression threads are
  spawned, matching the other numeric flags.
- Drop the unused ShardWriters::shard_number field and into_parts helper.
- Add tests for the out-of-sync orphan case and for surfacing a corrupt-gzip
  decompression error through the ByteChunker -> ChunkReader path.

tfenne committed:
docs: document shard and subsample in the README; generalize update-docs.sh

The README claimed fqtk had "a single tool, demux" and documented only demux;
shard (this PR) and subsample (#79) were absent from the README entirely, and
update-docs.sh only regenerated the demux usage block.

- Rewrite the intro to list all three tools with one-line descriptions.
- Add per-tool usage sections delimited by <!-- start/end usage:TOOL --> markers.
- Generalize update-docs.sh to loop over demux, shard, and subsample, regenerating
  each block from its --help output. Verified idempotent so the CI doc-sync check
  remains green.

tfenne committed:
Address CodeRabbit review: zero-length reads and error-path shutdown.

- ChunkReader::read returns Ok(0) immediately for an empty buffer, per the Read
  contract, instead of advancing the chunk iterator (which could block).
- Shard::execute captures the processing result and always runs the writer-pool
  shutdown (ShardWriters::close + Pool::stop_pool), even on error, so queued
  output is flushed and a finalization error is propagated rather than panicking
  in a drop. A processing error takes precedence over a cleanup error.
